### DISCUSSION TITLE
Helix, clangd, & OpenGL - and they aren't mixing.

### DISCUSSION THREAD

**Discussion by ghost** [clangd/clangd#2147] [Q&A]
Alright, I know the title may sound a bit vague, but here's the gist of my question.
Basically, I'm setting up OpenGL to work inside of Helix using a makefile; with clangd as a my lsp.
The main.c file compiles and runs fine (as far as i'm aware), but clangd is still giving me errors about not being able to find 'GLFW/glfw3/h' or 'glad/glad.h,' even those are already included in an _include_ folder that get linked.

![Screenshot 2024-09-17 215428](https://github.com/user-attachments/assets/85b44aa3-9ba2-4bb5-9410-6765f94476be)
Any thoughts on what's happening?

**Comment by HighCommander4** [comment]
I see this was already discussed on Discord, but for the benefit of anyone encountering this issue in the future: as discussed in https://clangd.llvm.org/installation#project-setup, you need to generate a `compile_commands.json` file to tell clangd how to understand your code.

For projects built using `make`, the tools most commonly used to generate `compile_commands.json` are https://github.com/rizsotto/Bear and https://github.com/nickdiego/compiledb; I believe only the latter works on Windows.

**Reply by ghost** [reply]
I tried to use compiledb, and it _did_ generate a .json file: but it was only filled with a pair of square brackets.
Is there a way to manually create the `compile_commands.json` file? I'm pretty sure there's also a `compile_flags.txt` that I could manually edit, but I haven't gotten any luck so far.

**Reply by HighCommander4** [reply]
> Is there a way to manually create the `compile_commands.json` file?

I suppose nothing stops you from reading the specification of the file format at https://clang.llvm.org/docs/JSONCompilationDatabase.html and manually authoring a file that conforms to it, but I wouldn't recommend it as that seems error prone.

For manual configuration, I would instead suggest using a [clangd config file](https://clangd.llvm.org/config.html) placed in the project's root directory. You can specify a set of compile flags that should be [added](https://clangd.llvm.org/config.html#add) to resolve include directories, for example:

```yaml
CompileFlags:
  Add: [
    -I/path/to/headers1,
    -I/path/to/headers2
  ]
```

There are some downsides to this compared to using `compile_commands.json`; in addition to the obvious one that you have to figure out and write down the include paths manually, a less obvious one is that without `compile_commands.json` clangd doesn't have a list of source files in the project to index, and as a result features that rely on the project index (such as finding references across all translation units) won't work. But it might still be useful as a temporary approach.

**Reply by i-ky** [reply]
> I tried to use compiledb, and it _did_ generate a .json file: but it was only filled with a pair of square brackets.

`compiledb` calls your `make` with certain options and then parses the output. Maybe your `make` outputs something `compiledb` does not expect. You can dive in, investigate and fix the parsing... but I would recommend swapping `make` for CMake that would generate `compile_commands.json` for you.
